Mathlib/Analysis/Calculus/Gradient/Lemmas.lean

```diff
@@ -0,0 +1,403 @@
+  simp only [sub_eq_add_neg, gradient_add_const]
+
+theorem HasGradientAtFilter.const_sub (hf : HasGradientAtFilter f f' x L) (c : 𝕜) :
+    HasGradientAtFilter (fun x => c - f x) (-f') x L := by
+  simpa only [sub_eq_add_neg] using hf.neg.const_add c
+
+theorem HasGradientWithinAt.const_sub (hf : HasGradientWithinAt f f' s x) (c : 𝕜) :
+    HasGradientWithinAt (fun x => c - f x) (-f') s x := by
+  exact HasGradientAtFilter.const_sub hf c
+
+theorem HasGradientAt.const_sub (hf : HasGradientAt f f' x) (c : 𝕜) :
+    HasGradientAt (fun x => c - f x) (-f') x := by
+  exact HasGradientAtFilter.const_sub hf c
+
+theorem gradient_const_sub (c : 𝕜) : ∇ (fun y => c - f y) x = - ∇ f x := by
+  calc
+    ∇ (fun y => c - f y) x = ∇ (fun y => - f y + c) x := by  congr; ext x; rw [sub_eq_neg_add]
+    _ = - ∇ f x := by rw [gradient_add_const c, gradient_neg]
+
+end Sub
+
+section Mul
+
+variable {a' b' c' d' : F} {a b c d : F → 𝕜} {x : F}
+
+open ContinuousLinearMap
+
+lemma equiv_lemma_mul : c x • (toDual 𝕜 F) d' + d x • (toDual 𝕜 F) c'
+    = (toDual 𝕜 F) ((starRingEnd 𝕜) (c x) • d' + (starRingEnd 𝕜) (d x) • c'):= by
+  simp only [map_add, map_smulₛₗ, RingHomCompTriple.comp_apply, RingHom.id_apply]
+
+theorem HasGradientAt.mul (hc : HasGradientAt c c' x) (hd : HasGradientAt d d' x) :
+    HasGradientAt (fun y => c y * d y)
+    ((starRingEnd 𝕜) (c x) • d' + (starRingEnd 𝕜) (d x) • c') x := by
+  rw [hasGradientAt_iff_hasFDerivAt, ← equiv_lemma_mul]
+  rw [hasGradientAt_iff_hasFDerivAt] at hc hd
+  exact hc.mul hd
+
+theorem HasGradientWithinAt.mul (hc : HasGradientWithinAt c c' s x)
+    (hd : HasGradientWithinAt d d' s x) :
+    HasGradientWithinAt (fun y => c y * d y) ((starRingEnd 𝕜) (c x) • d'
+      + (starRingEnd 𝕜) (d x) • c') s x := by
+  rw [hasGradientWithinAt_iff_hasFDerivWithinAt, ← equiv_lemma_mul]
+  rw [hasGradientWithinAt_iff_hasFDerivWithinAt] at hc hd
+  exact hc.mul hd
+
+theorem gradient_mul (hc : DifferentiableAt 𝕜 c x) (hd : DifferentiableAt 𝕜 d x) :
+    ∇ (fun y => c y * d y) x = (starRingEnd 𝕜) (c x) • ∇ d x
+      + (starRingEnd 𝕜) (d x) • ∇ c x :=
+  (hc.hasGradientAt.mul hd.hasGradientAt).gradient
+
+variable [InnerProductSpace ℝ F] {a' b' c' d' : F} {a b c d : F → ℝ} {x : F}
+
+theorem HasGradientAt.mul' (hc : HasGradientAt c c' x) (hd : HasGradientAt d d' x) :
+    HasGradientAt (fun y => c y * d y) ((c x) • d' + (d x) • c') x :=
+  HasGradientAt.mul hc hd
+
+theorem HasGradientWithinAt.mul' (hc : HasGradientWithinAt c c' s x)
+    (hd : HasGradientWithinAt d d' s x) :
+    HasGradientWithinAt (fun y => c y * d y) ((c x) • d' + (d x) • c') s x :=
+  hc.mul hd
+
+theorem gradient_mul' (hc : DifferentiableAt ℝ c x) (hd : DifferentiableAt ℝ d x) :
+    ∇ (fun y => c y * d y) x = (c x) • ∇ d x + (d x) • ∇ c x :=
+  gradient_mul hc hd
+
+end Mul
+
+section Mul_const
+
+variable {c' a': F} {c a: F → 𝕜} {x : F} (d b: 𝕜)
+
+lemma equiv_lemma_mul_const : d • (toDual 𝕜 F) c' = (toDual 𝕜 F) ((starRingEnd 𝕜) d • c') := by
```

Comment:
again, the lemma name is not good. But I would probably just remove it: it seems to me that all the lemmas in this section can be deduced from the lemmas of the previous one, together with the fact that a constant function has zero gradient, no? This would probably be more direct.